### PR TITLE
Endpoints to get all project types and get project with specified type

### LINKED ISSUE #17: Get all projects with the specified type
**Types of project:** 
Murale extérieure
Murale intérieure
Marquage au sol
Événement spécial
Murale éphémère
Mosaïque

An example – https://github.com/celinehuang/kom_get_your_bucha/blob/59b5261bbd4c02c2e0a918e8f3f43e1170b97323/backend/app/items/views.py#L23

### PR DESCRIPTION
**Pull request by peterghrong**
# Summary

1. Created endpoint to get all possible project types
2. Created endpoint to get Project object with specified project type
3. Did unit testing
	
## Test Plan

###Unit tested
1. Correct types returned by get_all_project_types()
2. Correct Project returned by get_project_of_type()
Screen Shot 2021-01-22 at 12 53 23 AM

## Related Issues

Which issues does this PR resolve/work on?
Closes #17

### CODE REVIEW

File: backend/requirements.txt [hunk 1]
```diff
@@ -23,3 +23,5 @@ toml==0.10.2
 typed-ast==1.4.1
 typing-extensions==3.7.4.3
 Werkzeug==1.0.1
+#added psycopg2-binary to run the test script
+psycopg2-binary==2.8.6
```
Comment (CurtisMIT) [inline backend/requirements.txt:27]: @peterghrong I was wondering whether you've tried it with just psycopg2-2.8.6? The issue you had before maybe caused by psycopg2-2.7.3 and I have since updated it back to 2.8.6. You can test it by removing your virtual environment directory and installing the dependencies again.

File: backend/requirements.txt [hunk 2]
```diff
@@ -13,7 +13,7 @@ Mako==1.1.3
 MarkupSafe==1.1.1
 mypy-extensions==0.4.3
 pathspec==0.8.0
-psycopg2==2.8.6
+# psycopg2==2.8.6
```
Comment (CurtisMIT) [inline backend/requirements.txt:16]: @peterghrong Why is this removed and then added again?

File: backend/tests/test_project.py [hunk 3]
```diff
@@ -71,3 +71,99 @@ def test_project_routes(self):
...
+            province="dummy province",
+            postal_code="dummy postal_code",
+            neighbourhood="dummy neighbourhood",
+            year=2020,
+            name="dummy name",
+            type=test_list[1],
+            contacts=[],
+        )
+        db.session.add(p1)
+        db.session.add(p2)
```
Comment (CurtisMIT) [inline backend/tests/test_project.py:100]: @peterghrong If we're going to add more than one, try using `add_all`
https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.add_all

File: backend/tests/test_project.py [hunk 4]
```diff
@@ -71,3 +71,99 @@ def test_project_routes(self):
...
+            postal_code="dummy postal_code",
+            neighbourhood="dummy neighbourhood",
+            year=2020,
+            name="dummy name",
+            type=test_list[0],
+            contacts=[],
+        )
+        db.session.add(p1)
+        db.session.add(p2)
+        db.session.add(p3)
```
Comment (CurtisMIT) [inline backend/tests/test_project.py:145]: @peterghrong Same here, https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.add_all

File: backend/tests/test_project.py [hunk 5]
```diff
@@ -71,3 +71,99 @@ def test_project_routes(self):
...
+            "postal_code": "dummy postal_code",
+            "neighbourhood": "dummy neighbourhood",
+            "year": 2020,
+            "name": "dummy name",
+            "type": test_list[1],
+            "contacts": [],
+        }
+        # test empty type param
+        response = self.client.get("/project/")
+        self.assertEqual(response.status_code, 404)
```
Comment (CurtisMIT) [inline backend/tests/test_project.py:161]: @peterghrong This might not return a 404 error code once #98 is merged as there is an actual route that makes it a valid request. I could be wrong here, I was wondering whether you could check up on this, or show that it will continue to return a 404 status code even if #98 is merged (a screenshot)?

Comment (peterghrong) [reply backend/tests/test_project.py:161]: Good point Curtis, thanks for pointing this out, I tested it and it seems that it works fine 
Screen Shot 2021-01-24 at 12 16 15 PM

Comment (CurtisMIT) [reply backend/tests/test_project.py:161]: @peterghrong Perfect, tysm!!!! 💯

File: backend/requirements.txt (hunk 1 above)
Comment (CurtisMIT) [reply backend/requirements.txt:27]: @peterghrong jk keeping both should capture everyone's situation. Mb!

File: backend/tests/test_project.py (hunk 5 above)
Comment (peterghrong) [reply backend/tests/test_project.py:161]: I tried to merge in the master branch and found that `/project/<uuid>` and `/project/<type>` have route conflict. I added a extra param into the url to resolve this so now we have: 
eg. 
- to get all types
`/project/type/types 
`
- to get project by uuid
`/project/id/<uuid>`

Fixed all unittests as well and ran them.
Screen Shot 2021-01-25 at 11 05 49 AM

**Review by CurtisMIT — CHANGES_REQUESTED** [review]
@peterghrong A few unresolved comments from last review and some new comments on changes other than the route names.

Regarding the naming conventions of api routes, nice job of pointing out the conflict. In general when identifying a specific instance of a resource (projects) by its id (uuid), we usually have it as a path variable without additional prefixes after the resource name (/projects/<uuid:id>). Whereas when querying, filtering or searching based on a parameter (type) we usually would prefer using query parameters (/projects/?type=type1). Again these are all conventions so there's not really a right or wrong, open for discussion. I've linked a few sites that I skimmed through, maybe of some help.

@celinehuang Now that I've taken another look, is the `type` in `project` the same as `contact_type` is to a `contact`? If so would we want to create another model for it specifically?

Links:
https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api
https://medium.com/@fullsour/when-should-you-use-path-variable-and-query-parameter-a346790e8a6d
https://www.moesif.com/blog/technical/api-design/REST-API-Design-Best-Practices-for-Sub-and-Nested-Resources/

**Comment by celinehuang** [comment]
hmm good point @CurtisMIT I believe the reason why I didn't think it was necessary to separate out project type from project because this was strictly 1:1 relationship whereas a contact could be multiple types. Although I should confirm this with Mu but for now, let's just keep it as a field in `project`

File: backend/tests/test_project.py [hunk 6]
```diff
@@ -99,21 +196,14 @@ def test_get_project_by_id(self):
         )
         p = Project(
             id=p_id,
-            address="dummy address",
-            city="dummy city",
-            province="dummy province",
-            postal_code="dummy postal_code",
-            neighbourhood="dummy neighbourhood",
-            year=2020,
-            name="dummy name",
```
[the remark this answers is not in the document]
Comment (peterghrong) [reply backend/tests/test_project.py:108]: Mb Added this back

**Comment by peterghrong** [comment]
> @peterghrong A few unresolved comments from last review and some new comments on changes other than the route names.
> 
> Regarding the naming conventions of api routes, nice job of pointing out the conflict. In general when identifying a specific instance of a resource (projects) by its id (uuid), we usually have it as a path variable without additional prefixes after the resource name (/projects/uuid:id). Whereas when querying, filtering or searching based on a parameter (type) we usually would prefer using query parameters (/projects/?type=type1). Again these are all conventions so there's not really a right or wrong, open for discussion. I've linked a few sites that I skimmed through, maybe of some help.
> 
> @celinehuang Now that I've taken another look, is the `type` in `project` the same as `contact_type` is to a `contact`? If so would we want to create another model for it specifically?
> 
> Links:
> https://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api
> https://medium.com/@fullsour/when-should-you-use-path-variable-and-query-parameter-a346790e8a6d
> https://www.moesif.com/blog/technical/api-design/REST-API-Design-Best-Practices-for-Sub-and-Nested-Resources/

Thanks for all these links Curtis, I've updated everything accordingly

File: backend/app/__init__.py [hunk 7]
```diff
@@ -53,4 +53,8 @@ def create_app(config_name):
 
     app.register_blueprint(project_blueprint, url_prefix="/project")
 
+    from .project import project as project_blueprint
+
+    app.register_blueprint(project_blueprint, url_prefix="/project")
```
[the remark this answers is not in the document]
Comment (CurtisMIT) [reply backend/app/__init__.py:58]: @peterghrong extra bump

File: backend/tests/test_project.py (hunk 3 above)
Comment (CurtisMIT) [reply backend/tests/test_project.py:100]: @peterghrong extra bump

File: backend/tests/test_project.py (hunk 4 above)
Comment (CurtisMIT) [reply backend/tests/test_project.py:145]: @peterghrong extra bump

File: backend/tests/test_project.py [hunk 8]
```diff
@@ -72,11 +72,108 @@ def test_project_routes(self):
...
+        response3 = self.client.get("/project?type={}".format(test_list[0]))
+        json_response = json.loads(response3.get_data(as_text=True))
+        self.assertEqual(len(json_response), 2)
+
     def test_get_project_by_id(self):
         p_id = uuid.uuid4()
 
         # get a project with an id that does not exist
         response = self.client.get("/project/{}".format(p_id))
+        json_response = response.get_json()
```
Comment (CurtisMIT) [inline backend/tests/test_project.py:176]: @peterghrong why is this added?

File: backend/tests/test_project.py [hunk 9]
```diff
@@ -99,15 +196,15 @@ def test_get_project_by_id(self):
         )
         p = Project(
             id=p_id,
+            type="dummy type",
+            contacts=[c],
```
Comment (CurtisMIT) [inline backend/tests/test_project.py:200]: @peterghrong minor issue, usually for places in the code where you made accidental changes, you would want to revert it back to how it was originally before. So there should not be any line changes in this part.

**Review by CurtisMIT — CHANGES_REQUESTED** [review]
@peterghrong The new changes regarding the route changes look 💯 . Besides that just a few missed comments from previous reviews, should be good after that.

**Comment by celinehuang** [comment]
> hmm good point @CurtisMIT I believe the reason why I didn't think it was necessary to separate out project type from project because this was strictly 1:1 relationship whereas a contact could have multiple types. Although I should confirm this with Mu but for now, let's just keep it as a field in `project`

Project only has 1 type confirmed, no need to add a separate model

File: backend/tests/test_project.py (hunk 9 above)
Comment (peterghrong) [reply backend/tests/test_project.py:196]: Will do next time

File: backend/tests/test_project.py [hunk 10]
```diff
@@ -99,15 +192,15 @@ def test_get_project_by_id(self):
...
```
Comment (CurtisMIT) [inline backend/tests/test_project.py:196]: @peterghrong Is it possible for you to revert it back as these are unnecessary changes.

File: backend/tests/test_project.py [hunk 11]
```diff
@@ -99,15 +192,15 @@ def test_get_project_by_id(self):
...
             address="dummy address",
             city="dummy city",
             province="dummy province",
             postal_code="dummy postal_code",
             neighbourhood="dummy neighbourhood",
             year=2020,
             name="dummy name",
-            type="dummy type",
-            contacts=[c],
```
Comment (CurtisMIT) [inline backend/tests/test_project.py:110]: @peterghrong same here.

**Review by CurtisMIT — CHANGES_REQUESTED** [review]
@peterghrong one last fix needed and should be good!

File: backend/app/project/routes.py [hunk 12]
```diff
@@ -1,9 +1,30 @@
-import uuid
```
Comment (peterghrong) [inline backend/app/project/routes.py:1]: This uuid import seems to be unused

**Review by CurtisMIT — APPROVED** [review]
CHAMP 🥇
